Issue 450  clintec spect data

niinimaki opened the pull request:
Added functionality to read in CLINTEC SPECT/CT data
`clintec_spect.py` contains 3 functions

- `read_clintec_CT_reconstruction` reads in CT reconstructions in Dicom form and returns an array of  Houndsfield units

- `linear_attenuation_map_from_HU_data` calls `read_clintec_CT_reconstruction` and converts HU to linear attenuation map and returns the corresponding array

- `read_clintec_raw_spect_data` reads the Dicom SPECT data and returns to and array

kohr-h commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+
+Investigation of the relationship between linear attenuation
+coefficients and CT Houndsfield units using radionuclides for SPECT
+S. Brown et al, Applied Radiation and Isotopes 66 (2008) 1206-1212
```
Put this in the [references file](https://github.com/odlgroup/odl/blob/5b5d5fd04347d175593acd4e8eef936aec03d289/doc/source/refs.rst) using our [citation style](https://github.com/odlgroup/odl/issues/279) (compare with the stuff in the file).

kohr-h commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+
+
+def read_clintec_CT_reconstruction(path):
+    """Reads the data in Dicom format and returns a HU CT volume.
```
Read
DICOM

kohr-h commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+
+    Parameters
+    ----------
+    path : a `path` to the folder where Dicom data is
```
```
path : string
    blabla
```

adler-j replied:
```
path : str
    blabla
```

kohr-h commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+import numpy as np
+import dicom
+import os
+from natsort import natsorted
```
Why do we need this?

niinimaki replied:
To identify numbers anywhere in a string and to sort them naturally, as [in here](https://github.com/odlgroup/odl/blob/issue-450__clintec-spect-data/odl/tomo/data/clintec_spect.py#L52).

But if there is another way to do this without `natsort`, that could be used instead.

adler-j replied:
There has to be another way, adding dependencies like this is not very good. Worst case we implement it ourselves. See [Stackoverflow](http://stackoverflow.com/questions/4836710/does-python-have-a-built-in-function-for-string-natural-sort) for a discussion.

niinimaki replied:
This has been replaced by

`for filename in sorted(fileList):`

Seems to work. So `natsort` is no longer needed.

kohr-h commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+    lstFilesDCM = []  # create an empty list
+    for dirName, subdirList, fileList in os.walk(path):
+        for filename in natsorted(fileList):
+            if ".dcm" in filename.lower():
```
`filename.lower().endswith('.dcm')`

kohr-h commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+                # check whether the file ends by .dcm
+                lstFilesDCM.append(os.path.join(dirName, filename))
+            else:
+                lstFilesDCM.append(os.path.join(dirName, filename))
```
Exactly the same in the `else` case?

niinimaki replied:
True, I'll remove this.

kohr-h commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+
+    # Load dimensions based on the number of rows, columns, and slices
+    shape = (int(dataset.Rows), int(dataset.Columns), len(lstFilesDCM))
+    data_array = np.zeros(shape, dtype=dataset.pixel_array.dtype)
```
`np.zeros_like(dataset.pixel_array)` should work?

kohr-h commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+        # read the file
+        ds = dicom.read_file(filenameDCM)
+        # store the raw image data
+        data_array[:, :, lstFilesDCM.index(filenameDCM)] = ds.pixel_array
```
Use
```python
for i, fname in enumerate(file_list):
    ds = dicom.read_file(fname)
    data_array[:, :, i] = ds.pixel_array
```

kohr-h commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+    Parameters
+    ----------
+    path : a `path` to the folder where Dicom data is
+
```
Returns?

kohr-h commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+    return ct_volume, dataset
+
+
+def linear_attenuation_map_from_HU_data(path, a, b):
```
Better not call it data, it has the connotation of projection data. `linear_attenuation_map_from_HU` should suffice.

kohr-h commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+    ----------
+    path : a `path` to the data folder
+    a, b : `tuple`
+        a = (a1, a2) where a1 `float` and a2 `float` are the intercepts
```
sequence

kohr-h commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+        a and b are parameters of a bilinear curve for converting
+        HU-values to linear attenuation map, such that
+        mu = a1 + b1 * ct_volume for HU < 0
+        mu = a2 + b2 * ct_volume for HU >= 0
```
Put this more detailed explanation into a `Notes` section.

kohr-h commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+        a = (a1, a2) where a1 `float` and a2 `float` are the intercepts
+            of the bilinear curve
+        b = (b1, b2) where b1 `float` and b2 `float` are the slopes
+            of the bilinear curve
```
Point to the (to be created) Notes section, otherwise this is just jargon for users.

kohr-h commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+    a1, a2 = a
+    b1, b2 = b
+
+    ct_volume, dataset = read_clintec_CT_reconstruction(path)
```
Don't intertwine this function with data reading. Just let it convert an existing array(-like).

kohr-h commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+    dataset = dicom.read_file(data_dicom)
+    spect_data = np.float32(dataset.pixel_array)
+
+    return spect_data
```
This cannot be all. There must be geometry information stored in the files that we need to read as well.

niinimaki replied:
That is true and I have modified this such that it also returns the `dataset`.

kohr-h commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+    filenames = [f for f in os.listdir(path)]
+    data_dicom = os.path.join(path, filenames[0])
+    dataset = dicom.read_file(data_dicom)
+    spect_data = np.float32(dataset.pixel_array)
```
Why do we convert to `float32`?

niinimaki replied:
I'll remove this.

kohr-h commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+    b1, b2 = b
+
+    ct_volume, dataset = read_clintec_CT_reconstruction(path)
+    (i1, j1, k1) = np.where(ct_volume < 0)
```
We don't need to get the index arrays. Just use
```python
negative = np.negative(ct_volume)  # More robust, allows array-like
attenuation_map[negative] = a1 + b1 * ct_volume[negative]
nonnegative = ~negative
attenuation_map[nonnegative] = a2 + b2 * ct_volume[nonnegative]
```

kohr-h replied:
Here comes the even more memory-saving version 2:
```python
mask = np.negative(ct_volume)  # More robust, allows array-like
attenuation_map[mask] = a1 + b1 * ct_volume[mask]
np.logical_not(mask, out=mask)
attenuation_map[mask] = a2 + b2 * ct_volume[mask]
```

adler-j replied:
Have you tried this code? Negative is not "is negative" but "the negative of". I suppose you want to write `mask = np.less_equal(a, 0)`?

kohr-h replied:
You're right. I tried it with a trivial example which happened to work :-)

kohr-h commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+
+    attenuation_map = np.zeros_like(ct_volume)
+    attenuation_map = np.zeros_like(ct_volume)
+    attenuation_map[i1, j1, k1] = a1 + b1 * ct_volume[i1, j1, k1]
```
We should consider avoiding temporary arrays, the volumes can be large. To use in-place arithmetic, this can be split into (see above comment)
```python
attenuation_map[negative] = ct_volume[negative]
attenuation_map[negative] *= b1
attenuation_map[negative] += a1
```
This code runs in fact a bit slower than the original one due to the multiple slicing. However, it's more memory friendly, so I would prefer it.

niinimaki replied:
This does not work
`negative = np.negative(ct_volume)`
`attenuation_map[negative] = ct_volume[negative]`

Doesn't `np.negative` take the element-wise numerical negatives of the values, but I would like to get the indices in the array for the negative (&positive) values...

adler-j replied:
This creates copys anyway since `negative` is not a slice, but an indicator array. To be "very efficient" here, I'd go for one copy:

```python
mask = np.less(ct_volume, 0)

attenuation_map = np.empty_like(ct_volume)

negative_part = ct_volume[mask]

negative_part *= b1
negative_part += a1

attenuation_map[mask] = negative_part

mask = np.logical_not(mask, out=mask)
positive_part = ct_volume[mask]

positive_part *= b2
positive_part += a2

attenuation_map[mask] = positive_part
```

With that said, I think we shouldnt overoptimize this at this point. The straightforward solution

```python
mask = np.less(ct_volume, 0)
attenuation_map = np.empty_like(ct_volume)
attenuation_map[mask] = a1 + b1 * ct_volume[mask]
attenuation_map[mask] = a2 + b2 * ct_volume[~mask]
```

should do fine.

kohr-h commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+def read_clintec_raw_spect_data(path):
+    """Reads the raw spect Dicom data and retuns an array of the data."""
+
+    filenames = [f for f in os.listdir(path)]
```
`list(os.listdir(path))` should do.

kohr-h commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+    """Reads the raw spect Dicom data and retuns an array of the data."""
+
+    filenames = [f for f in os.listdir(path)]
+    data_dicom = os.path.join(path, filenames[0])
```
What is special about the first entry?

niinimaki replied:
Contains the DICOM data, in the case of SPECT this list actually consists of one entry.

kohr-h replied:
I was guessing something like that. Can you write a comment?

niinimaki replied:
With respect to this I think that what @adler-j  wrote 

> Shouldnt we do an assert then that filenames has exactly one element? This feels like a bug waiting to happen.

might be a good idea...

kohr-h commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+standard_library.install_aliases()
+
+import numpy as np
+import dicom
```
This will fail on default installations since the `dicom` dependency is missing. Add this to the `setup.py` `extra_dependencies` under a name like `dicom`.

adler-j replied:
Even after adding it to `extra_dependencies` this will cause a fail. We need to have an include guard here, the classical

```python
try:
    import dicom
    DICOM_AVAILABLE = True
except ImportError:
    DICOM_AVAILABLE = False
```

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+
+    """
+
+    lstFilesDCM = []  # create an empty list
```
Also `# create an empty list` is overdocumentation, that is quite obvious from the code. If you want to write doc here, it would be to explain *why* the list is created.

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,120 @@
+    (i2, j2, k2) = np.where(ct_volume >= 0)
+
+    attenuation_map = np.zeros_like(ct_volume)
+    attenuation_map = np.zeros_like(ct_volume)
```
duplicate

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,129 @@
+
+    Notes
+    -----
+        a and b are parameters of a bilinear curve for converting
```
bilinear is perhaps not the correct word here

niinimaki replied:
bilinear is used for example in `[Bro+2008]` to which I refer.
There are two linear curves a one for HU < 0 and another for HU >= 0.
So piecewise linear would work also but bilinear seems to be used in the emission tomography literature.

adler-j replied:
Suppose we stay with that then!

niinimaki commented:
Revision of `clintec_spect`:

- moved the reference from `clintec_spect` into `refs.rst`

- added dicom to `setup` as `extra_require`

- improved the docstrings in `clintec_spect`

- improved the code of `clintec_spect` according to suggestions

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,129 @@
+
+    dcm_file_list = []
+    for dirName, subdirList, fileList in os.walk(path):
+        for filename in sorted(fileList):
```
This could be in the "wrong" order:

```python
>>> sorted(['a1', 'a2', 'a10', 'a11', 'a20'])
['a1', 'a10', 'a11', 'a2', 'a20']
```

I suppose that was why you used `natsort`?

niinimaki replied:
Yes. Is there another options for this not using `natsort`?

adler-j replied:
Yes several, e.g. [this](http://stackoverflow.com/questions/2545532/python-analog-of-natsort-function-sort-a-list-using-a-natural-order-algorithm).

niinimaki replied:
The options of not using `natsort` need `re`, wouldn't this be just another dependency?

adler-j replied:
`re` is a core package so that is fine, kindoff like using `os` or `sys`.

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,129 @@
+        ds = dicom.read_file(fname)
+        data_array[:, :, i] = ds.pixel_array
+
+    rescale_intercept = int(dataset.RescaleIntercept)
```
why `int`?

niinimaki replied:
should be `float` instead I think.

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,129 @@
+
+For the bilinear relationship of linear attenuation coefficients and
+Houndsfield units see [Bro+2008]_ and references therein, for example.
+
```
Remove blank line

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,129 @@
+    Parameters
+    ----------
+    path : `str`
+        a path to the folder where DICOM data is
```
`Path of the folder where DICOM data is stored`

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,129 @@
+    dataset = dicom.read_file(dcm_file_list[0])
+
+    shape = (int(dataset.Rows), int(dataset.Columns), len(dcm_file_list))
+    data_array = np.zeros(shape, dtype=dataset.pixel_array.dtype)
```
`np.empty`

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,129 @@
+
+    dataset = dicom.read_file(dcm_file_list[0])
+
+    shape = (int(dataset.Rows), int(dataset.Columns), len(dcm_file_list))
```
We usually store the angles in the first index, why use the last here?

niinimaki replied:
No strong reason, I'll change such that angles are first.

adler-j replied:
Short edit here: if this reads in a volume, we use the x, y, z convention, so then it should be last. Perhaps I missed that.

kohr-h replied:
I was about to write the same thing, but if I understood correctly, this is for reading a reconstruction by axial slices, so the length should come last.

adler-j commented on setup.py:
```diff
@@ -137,8 +137,9 @@ def run_tests(self):
-        'scikit' : 'scikit-image',
-        'all': test_requires + ['matplotlib', 'pyfftw', 'Pywavelets', 'scikit-image']
+        'scikit': 'scikit-image',
+        'dicom': 'dicom',
```
Are we sure this is not `pydicom`?

```bash
$ pip install dicom
...
No matching distribution found for dicom
$ pip install pydicom
...
Installing collected packages: pydicom
```

niinimaki replied:
One installs `pydicom` but imports it as [`import dicom`](http://pydicom.readthedocs.io/en/stable/getting_started.html).

So maybe this should be 
`'pydicom': 'dicom'`
or?

adler-j replied:
Other way around. To the left of the colon is our name for it, to the right is the pip package name.

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,129 @@
+    """
+
+    filenames = list(os.listdir(path))
+    data_dicom = os.path.join(path, filenames[0])
```
How do we know it is the first one? What is the ordering convention of `os.listdir`?

niinimaki replied:
As far as I know spect data in DICOM format is just one file, hence it first and only entry in the list...

adler-j replied:
Shouldnt we do an assert then that `filenames` has exactly one element? This feels like a bug waiting to happen.

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,129 @@
+    -------
+    spect_data : `numpy.ndarray`
+        an array of SPECT projection data
+    dataset : `DICOM` File Meta Information
```
is `DICOM` a dead link?

niinimaki replied:
I'll remove the ticks.

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,129 @@
+
+    Parameters
+    ----------
+    volume : `numpy.ndarray`
```
`array-like`

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,129 @@
+
+    for i, fname in enumerate(dcm_file_list):
+        ds = dicom.read_file(fname)
+        data_array[:, :, i] = ds.pixel_array
```
My old comment got removed here, but why do we store the "angles" in the last index and not the first like we do in `RayTransform`?

niinimaki replied:
> We usually store the angles in the first index, why use the last here?

> Short` edit here: if this reads in a volume, we use the x, y, z convention, so then it should be last. Perhaps I missed that.

Is this not the same comment as above?? under line 58??

adler-j replied:
Yes, just me being confused.

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,129 @@
+    """
+
+    dcm_file_list = []
+    for dirName, subdirList, fileList in os.walk(path):
```
`subdirList` unused, replace with `_`:

```
for dirName, _, fileList in os.walk(path):
```

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,129 @@
+    Returns
+    -------
+    ct_volume : `numpy.ndarray`
+        an array of CT reconstruction in Houndsfield units.
```
`Array of CT reconstruction in Houndsfield units.`

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,129 @@
+import os
+
+
+def read_clintec_CT_reconstruction(path):
```
In what way is this specific to `clintec`? It looks exactly like the MAYO data to me.

niinimaki replied:
Well if they really are same, maybe we then should have one functions that does this, the name of which not referring to any specific clinic..

adler-j replied:
Perhaps keep this for now and I can merge this when we fix the MAYO data.

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,129 @@
+    rescale_intercept = int(dataset.RescaleIntercept)
+    rescale_slope = int(dataset.RescaleSlope)
+
+    ct_volume = rescale_slope * np.float32(data_array) + rescale_intercept
```
`np.float32(data_array)` should be redundant?

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,129 @@
+            of the bilinear curve
+        b = (b1, b2) where b1 `float` and b2 `float` are the slopes
+            of the bilinear curve
+        see `Notes` for furher information
```
Is `Notes` a link here?

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,129 @@
+    Returns
+    -------
+    attenuation_map : `numpy.ndarray`
+        An array of linear attenuation values.
```
~~An~~

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,129 @@
+
+
+def read_clintec_raw_spect_data(path):
+    """Reads the raw spect DICOM data and retuns an array of the data.
```
`Read raw SPECT DICOM data to array.`

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,129 @@
+    Returns
+    -------
+    spect_data : `numpy.ndarray`
+        an array of SPECT projection data
```
`Array of SPECT projection data.`

niinimaki replied:
If I understand what you mean, aren't these data/scanner dependent??

adler-j replied:
Well, in general you want these types of methods to be quite specific in what they output, but ofc it depends on the call patterns. 

Thats why I feel like we need to get some code that actually calls this into ODL before we merge. An example or similar.

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,129 @@
+
+
+def linear_attenuation_map_from_HU(volume, a, b):
+    """Converts Houndsfield units to a linear attenuation map.
```
Convert~~s~~ Houndsfield units to linear attenuation coefficent.

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,129 @@
+    Parameters
+    ----------
+    volume : `numpy.ndarray`
+        an array of Houndsfield units to be converted
```
~~an~~

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,129 @@
+    return ct_volume, dataset
+
+
+def linear_attenuation_map_from_HU(volume, a, b):
```
Why do we use "linear attenuation **map**" but not ""HU **map**", I feel like we should be consistent here.

adler-j commented:
Review done, mostly minor stuff.

Do we have any code that can use the output of this, or is it simply dangling code?

adler-j commented on setup.py:
```diff
@@ -137,8 +137,9 @@ def run_tests(self):
-        'all': test_requires + ['matplotlib', 'pyfftw', 'Pywavelets', 'scikit-image']
+        'scikit': 'scikit-image',
+        'dicom': 'dicom',
+        'all': test_requires + ['matplotlib', 'pyfftw', 'Pywavelets', 'scikit-image', 'dicom']
```
`dicom` -> `pydicom`

niinimaki commented:
> Do we have any code that can use the output of this, or is it simply dangling code?

Not yet, except that `linear_attenuation_map_from_HU` can be used to get the `attenuation` for `spect_trafo`, provided that one has a volume in HU...

niinimaki commented:
Added one more reference related to `linear_attenuation_from_HU`.
Improved further the docstrings and code of `clintec_spect` according to suggestions.

kohr-h commented on doc/source/refs.rst:
```diff
@@ -16,12 +16,23 @@ References
+.. [Bro+2008] Brown, S, and Bailey, D L, and Willowson, K, and Baldock, C
+   *Investigation of the relationship between linear attenuation
+   coefficients and CT Houndsfield units using radionuclides for SPECT*.
+   Applied Radiation and Isotopes, 66 (2008), pp 1206--1212
```
Almost :-) Final dot missing.

kohr-h commented on doc/source/refs.rst:
```diff
@@ -16,12 +16,23 @@ References
+   and Lonn, A H R, and von Schulthess, G K *PET attenuation coefficient
+   from CT images: experimental evaluation of the transformation of CT into
+   PET 511-keV attenuation coefficients*.
+   European Journal of Nuclear Medicine, 29 (2002), pp 922--927
```
`[Bur+2002]` and final dot missing.

niinimaki commented:
Fixed the missing dot etc in the `refs.rst` related to `clintec_spect`.

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,152 @@
+        dataset = dicom.read_file(data_dicom)
+        spect_data = dataset.pixel_array
+    else:
+        raise NotImplementedError('folder contais more than one item')
```
I'd go for `ValueError` here.

kohr-h replied:
If it's always one file, why don't we simply give the filename instead of a folder?

niinimaki replied:
That could actually be a good idea.

adler-j commented on odl/tomo/data/clintec_spect.py:
```diff
@@ -0,0 +1,152 @@
+    attenuation = np.empty_like(volume)
+    attenuation[mask] = a1 + b1 * volume[mask]
+    attenuation[~mask] = a2 + b2 * volume[~mask]
+    attenuation = np.absolute(attenuation)
```
This step is not mentioned in the documentation above and could come as a surprise to users.

niinimaki replied:
This step is to ensure that there are no negative values in the attenuation coefficients (as there should not be). Negative values can occur due to the fact that the parameters `a` and `b` are approximations (evaluated for example for different scanners) and contain some errors them selves.  
But I see your point.

adler-j commented:
Starting to look very good by now. My only major concern is that this is dangling code with no examples nor tests and no obvious way of using it.

What use-case do we intend it for, can we somehow merge this into that branch instead of master?

niinimaki commented:
Making a new PR into the SPECT trafo branch.
